src/Development/Rattle/Hash.hs

```diff
@@ -24,65 +25,99 @@ import System.FilePath
 import Numeric
 import Control.Exception.Extra
 import Control.DeepSeq
+import GHC.Generics
+import Data.Serialize
+import General.FileName
+import General.FileInfo
 
+newtype Hash = Hash BS.ByteString
+    deriving (NFData, Show, Read, Eq, Hashable, Generic)
 
-newtype Hash = Hash String
-    deriving (NFData, Show, Read, Eq, Hashable)
-
+instance Serialize Hash
 
 mkHash :: BS.ByteString -> Hash
-mkHash = Hash . concatMap (f . ord) . BS.unpack
-    where f i = ['0' | i < 16] ++ showHex i ""
-
+mkHash = Hash
 
 -- Hashing lots of files is expensive, so we keep a cache
 {-# NOINLINE hashCache #-}
-hashCache :: IORef (Map.HashMap FilePath (UTCTime, Hash))
+hashCache :: IORef (Map.HashMap FileName (ModTime, Hash))
 hashCache = unsafePerformIO $ newIORef Map.empty
 
-
-getModTime :: FilePath -> IO (Maybe UTCTime)
-getModTime x = handleBool isDoesNotExistError (const $ return Nothing) (Just <$> getModificationTime x)
-
-toHashForward :: FilePath -> Maybe FilePath
-toHashForward x | ".rattle.hash" `isSuffixOf` x = Nothing
-                | otherwise = Just $ x <.> "rattle.hash"
-
-fromHashForward :: FilePath -> Maybe FilePath
-fromHashForward x | Just x <- stripSuffix ".rattle.hash" x = Just x
-                  | otherwise = Nothing
-
+toHashForward :: FileName -> Maybe FileName
+toHashForward x = let b = fileNameToByteString x
+                      s = BS.pack ".rattle.hash" in
+                    if BS.isSuffixOf s b then Nothing
+                    else Just $ fileNameFromByteString $ BS.append b s
+                    
+fromHashForward :: FileName -> Maybe FileName
+fromHashForward x = let b = fileNameToByteString x
+                        s = BS.pack ".rattle.hash" in
+                      fileNameFromByteString <$> BS.stripSuffix s b
+
+hashFileForwardIfStale :: FileName -> ModTime -> Hash -> IO (Maybe Hash)
+hashFileForwardIfStale file mt h =
+  case toHashForward file of
+    Nothing -> hashFileIfStale file mt h
+    Just file2 -> do
+      b2 <- doesFileNameExist file2
+      if not b2 then hashFileIfStale file mt h else do
+        b <- doesFileNameExist file
+        if not b then return Nothing else hashFileIfStale file2 mt h
+
+hashFileIfStale :: FileName -> ModTime -> Hash -> IO (Maybe Hash)
+hashFileIfStale file mt h = do
+  start <- getModTime file
+  case start of
+    Nothing -> return Nothing
+    Just start -> do
+      mp <- readIORef hashCache
+      case Map.lookup file mp of
+        Just (time,hash) | time == start -> return $ Just hash
+        _ -> if start == mt
```

Comment:
Easier as a guard - `_ | start == mt -> do f start h; return $ Just h`, and then each set of clauses can stand alone.